pyccel/codegen/wrapper/fortran_to_c_wrapper.py

```diff
@@ -315,59 +262,149 @@ def _wrap_Interface(self, expr):
-        are necessary, however these objects are created by the `BindCFunctionDefArgument`
-        class.
+        Extract the C-compatible FunctionDefArgument from the Fortran FunctionDefArgument.
 
-        The objects which describe the argument passed to the `expr` argument of the
-        original function are also created here. However the expressions necessary to
-        collect the information from the BindCFunctionDefArgument in order to create
-        these objects are left for later. This is done to ensure that optionals are
-        handled locally to the function call. This ensures that we do not duplicate if
-        conditions.
+        The extraction is done by finding the appropriate function
+        _extract_X_FunctionDefArgument for the object expr. X is the class type of the
+        variable stored in the object expr. If this function does not exist then the
+        method resolution order is used to search for other compatible
+        _extract_X_FunctionDefArgument functions. If none are found then an error is raised.
 
         Parameters
         ----------
         expr : FunctionDefArgument
-            The argument to be wrapped.
+            An object representing the FunctionDefArgument in the Fortran code which should
+            be exposed to the C code.
+
+        func : FunctionDef
+            The function being wrapped.
 
         Returns
         -------
-        BindCFunctionDefArgument
-            The C-compatible argument.
+        dict
+            A dictionary describing the objects necessary to access the argument.
         """
         var = expr.var
+        class_type = var.class_type
+
+        classes = type(class_type).__mro__
+        for cls in classes:
+            annotation_method = f'_extract_{cls.__name__}_FunctionDefArgument'
+            if hasattr(self, annotation_method):
+                func_def_argument_dict = getattr(self, annotation_method)(var, func)
+                new_var = func_def_argument_dict['c_arg']
+                func_def_argument_dict['c_arg'] = FunctionDefArgument(new_var, value = expr.value,
+                    kwonly = expr.is_kwonly, annotation = expr.annotation,
+                    bound_argument = expr.bound_argument, persistent_target = expr.persistent_target)
+                func_def_argument_dict['f_arg'] = FunctionCallArgument(func_def_argument_dict['f_arg'],
+                                                                        keyword = expr.name)
+                return func_def_argument_dict
+
+        # Unknown object, we raise an error.
+        return errors.report(f"Wrapping function arguments is not implemented for type {class_type}. "
+                + PYCCEL_RESTRICTION_TODO, symbol=var, severity='fatal')
+
+    def _extract_FixedSizeNumericType_FunctionDefArgument(self, var, func):
         name = var.name
         self.scope.insert_symbol(name)
         collisionless_name = self.scope.get_expected_name(var.name)
```

Comment:
`var.name` can be replaced with `name`

Reply:
Fixed [f9365af](https://github.com/pyccel/pyccel/pull/2232/commits/f9365af03c0482c7585c3d4288b8df0b6307a067)